webrtc/src/rtp_transceiver/srtp_writer_future.rs

```diff
@@ -12,6 +12,92 @@ use interceptor::{Attributes, RTCPReader, RTPWriter};
 use std::sync::atomic::{AtomicBool, Ordering};
 use std::sync::{Arc, Weak};
 use tokio::sync::Mutex;
+use util;
+
+/// `RTP` packet sequence number manager.
+///
+/// Used to override outgoing `RTP` packets' sequence numbers. On creating it is
+/// unabled and can be enabled before sending data begining. Once data sending
+/// began it can not be enabled any more.
+pub(crate) struct SequenceTransformer(util::sync::Mutex<SequenceTransformerInner>);
+
+/// [`SequenceTransformer`] inner.
+struct SequenceTransformerInner {
+    offset: u16,
+    last_sq: u16,
+    reset_needed: bool,
+    enabled: bool,
+    data_sent: bool,
+}
+
+impl SequenceTransformer {
+    /// Creates a new [`SequenceTransformer`].
+    pub(crate) fn new() -> Self {
+        Self(util::sync::Mutex::new(SequenceTransformerInner {
+            offset: 0,
+            last_sq: rand::random(),
+            reset_needed: false,
+            enabled: false,
+            data_sent: false,
+        }))
+    }
+
+    /// Enables this [`SequenceTransformer`].
+    ///
+    /// # Errors
+    ///
+    /// With [`Error::ErrRTPSenderSeqTransEnabled`] on trying to enable already
+    /// enabled [`SequenceTransformer`].
+    ///
+    /// With [`Error::ErrRTPSenderSeqTransEnabled`] on trying to enable
+    /// [`SequenceTransformer`] after data sending began.
+    pub(crate) fn enable(&self) -> Result<()> {
+        let mut guard = self.0.lock();
+
+        if guard.enabled {
+            return Err(Error::ErrRTPSenderSeqTransEnabled);
+        }
+
+        (!guard.data_sent)
+            .then(|| {
+                guard.enabled = true;
+            })
+            .ok_or(Error::ErrRTPSenderDataSent)
+    }
+
+    /// Indicates [`SequenceTransformer`] about necessity of recalculating
+    /// `offset`.
+    pub(crate) fn reset_offset(&self) {
+        self.0.lock().reset_needed = true;
+    }
+
+    /// Gets [`Some`] consistent `sequence number` if this [`SequenceTransformer`] is
+    /// enabled or [`None`] if it is not.
+    ///
+    /// Once this method is called, considers data sending began.
+    fn seq_number(&self, raw_sn: u16) -> Option<u16> {
+        let mut guard = self.0.lock();
+        guard.data_sent = true;
```

Comment:
Shouldn't this be after `guard.enabled` below?

Reply:
No, it shouldn't.
We have already discussed it with you in discord. If data is sent we are not allowed to enable `SequenceTransformer`.

Reply:
So, we should register data sending even if `SequenceTransformer` is not `enabled`.

Reply:
Ah, right I recall